An option to write the result to a new file.

benjyw opened the pull request:
Useful in hermetic execution sandboxes in which input files can't be modified by tools in the sandbox.

This issue has been seen in the wild in an implementation of the REAPI, which writes the input files as read-only,
and which appears to be correct behavior per the spec (e.g., if the sandbox inputs were assembled by linking
to files in a loose-file CAS implementation).

benjyw committed:
An option to write the result to a new file.

Useful in hermetic execution sandboxes in which input
files can't be modified by tools in the sandbox.

stuhood reviewed (approved):
Looks good: thanks!

jsirois commented:
I'd prefer no more work on Lambdex and instead folks just use PEX + the magic `__pex__` package import prefix since that has been working now for a while. If there is a rush and that's prohibitive I guess that's one thing though. Is there?

benjyw commented:
Yeah, unfortunately this is a bug that is blocking upgrades to Pants 2.15.0rc3 if you're using remote execution.

I'd also like to get rid of Lambdex, but that's a somewhat bigger piece of work (albeit not huge either) that I don't want to do in 2.15 as we're stabilizing it.

jsirois commented:
It's confusing how a Pants (client) version could somehow be tied up with a - presumably - purely server side remote execution error. I'd think this would be an error in ~any Pants version.

Is there an associated Pants issue?

benjyw committed:
Ensure file permissions are preserved

jsirois commented on lambdex/bin/lambdex.py:
```diff
@@ -56,6 +57,13 @@ def write_lambdex_handler(pex_zip, options):
+    if options.output:
+        output_zip = options.output
+        shutil.copy(pex_zip, output_zip, follow_symlinks=False)
+        output_zip = options.script
```
Is this line intentional?

benjyw replied:
No that was a dumb autocomplete typo while copying that line...

benjyw committed:
Fix dumb typo bug

benjyw commented:
> It's confusing how a Pants (client) version could somehow be tied up with a - presumably - purely server side remote execution error. I'd think this would be an error in ~any Pants version.
> 
This is because we only enabled environments for the awslambda plugin (and many others) in https://github.com/pantsbuild/pants/pull/18144

benjyw committed:
Don't use follow_symlinks

The default is correct (and my setting it was not, I misread the
meaning), and also it's not available in 2.7.

jsirois commented:
Ok, but this presumably has ~nothing to do with environments. I.E. just with remexec (which environments can activate IIUC, but don't have to). If that's right, this is a bug for Pants + Lambdex + remexec going back ~infinity.

benjyw commented:
Yes, it is going back infinity, but was not triggered in the wild (that we know of) until attempting to use remexec in Toolchain's CI, which now attempted to remote this process due to environments. I will open an issue over in pantsbuild/pants so that this is documented.

benjyw committed:
Tox allowlist

benjyw committed:
Dammit tox
